### PR TITLE
Make json-rpc/call effect more in line with re-frame standards

### PR DESCRIPTION
**Pull request by ilmotta**
## Summary

This PR changes `:json/rpc-call` effect to accept `on-success` and `on-error` parameters either as functions OR as re-frame *event vectors*. The changes are 100% backwards compatible.

Note 1: This is one of those PRs that take longer to explain, than to actually write the code ;)
Note 2: Added lots of contributors for the sake of visibility.

## Why?

Re-frame is very explicit in its documentation and its architecture, saying that event handlers should be pure. Calling `re-frame.core/dispatch` in event handlers makes them stateful.

> So, you can "get away with it". But it ain't pure.
>
> -- https://day8.github.io/re-frame/EffectfulHandlers/#90-solution

Okay, but maybe you are thinking I'm being too *puristic*, and following re-frame's recommendation doesn't give us any real benefit.

In `status-mobile`, arguably one of our most important *effects* (not to be confused with event handlers) is `:json-rpc/call`, but at the moment, the `on-success` and `on-error` values are expected to be stateful functions (e.g. usually used for logging and dispatching subsequent events).

This creates two important problems:

1. The value returned by event handlers is more opaque and cannot be easily inspected (for example using `tap>`, `log/debug` or just `println`). If we try to inspect or log them, `on-success` and `on-error` will be printed as `#object[Function]`.
2. Testing event handlers using `:json-rpc/call` becomes an exercise of frustration, because we can't simply compare the results of the event handler with a good *expected* value, which is one of the big selling points of testing pure functions.

### The testability of event handlers

> For any re-frame app, there's three things to test:
>
> - Event Handlers - most of your testing focus will be here because this is
>   where most of the logic lives
> - Subscription Handlers - often not a lot to test here. Only Layer 3
>   subscriptions need testing.
> - View functions - I don't tend to write tests for views.
>
> -- https://day8.github.io/re-frame/Testing/#what-to-test

So re-frame is saying event handlers should be pure, and that event handlers should be tested.

In order to achieve the divine simplicity of testing event handlers as pure functions, we need to make `:json-rpc/call` *also* accept `on-success` and `on-error` as event vectors.

Good news is that there is a known pattern to solve this problem, e.g. used by the library https://github.com/Day8/re-frame-http-fx.

The pattern is simple once we see it: behind the scenes, `:json-rpc/call` `conj`es the results of the RPC call into the event vectors `on-success` and `on-error`, and `:json-rpc/call` dispatches the events itself. This eliminates the need for the stateful `dispatch` call in event handlers.

#### Example before/after

From this PR onward, we can **optionally** describe the `:json-rpc/call` effect using just data. It would be nice if all usages in the codebase used only the new approach, but our project is too big for me to tackle the full rewrite.

```clojure
;; BEFORE
(rf/defn delete-notification
  {:events [:activity-center.notifications/delete]}
  [{:keys [db]} notification-id]
  {:json-rpc/call [{:method     "wakuext_deleteActivityCenterNotifications"
                    :params     [[notification-id]]
                    :on-success #(rf/dispatch [:activity-center.notifications/delete-success notification-id])
                    :on-error   #(rf/dispatch [:activity-center/process-notification-failure notification-id :notification/delete %])}]})

;; AFTER
(rf/defn delete-notification
  {:events [:activity-center.notifications/delete]}
  [{:keys [db]} notification-id]
  {:json-rpc/call [{:method     "wakuext_deleteActivityCenterNotifications"
                    :params     [[notification-id]]
                    :on-success [:activity-center.notifications/delete-success notification-id]
                    :on-error   [:activity-center/process-notification-failure notification-id :notification/delete]}]})
```

## What about other effects?

There are other effects that are not cleanly exposing a data-only interface to event handlers, but I think this is a solid first step.

status: ready

### CODE REVIEW

File: src/status_im2/contexts/activity_center/events_test.cljs [hunk 1]
```diff
@@ -1,163 +1,169 @@
 (ns status-im2.contexts.activity-center.events-test
```
Comment (ilmotta) [inline src/status_im2/contexts/activity_center/events_test.cljs:1]: FYI: The diff in this file is somewhat big. The important point is that I was able to dramatically simplify them due to the simplicity of using `(is (= expected actual`) now that `:json-rpc/call` is just data in the file `activity_center/events.cljs`.

**Review by cammellos — APPROVED** [review]
Looks good!
The only thing worth considering, if that's the consensus going forward (vectors instead of fn), maybe you want to print a warning if a fn is passed instead, logging the event handler? so we can bit by bit identify and replace those as well, but up to us, no strong opinions on the matter

**Review by cammellos — APPROVED** [review]
Looks good!
The only thing worth considering, if that's the consensus going forward (vectors instead of fn), maybe you want to print a warning if a fn is passed instead, logging the event handler? so we can bit by bit identify and replace those as well, but up to us, no strong opinions on the matter

**Review by rahulpsingh — APPROVED** [review]

**Comment by ilmotta** [comment]
> The only thing worth considering, if that's the consensus going forward (vectors instead of fn), maybe you want to print a warning if a fn is passed instead, logging the event handler? so we can bit by bit identify and replace those as well, but up to us, no strong opinions on the matter

It's a good idea, @cammellos! The only thing is that, at this point, there would be too many warnings. So, as a starting point, we should probably log at lowish level, e.g. debug. But, as you said, let's see how much consensus there is first.

**Review by briansztamfater — APPROVED** [review]
Looks good, and appreciate the detailed explanation :)

**Review by smohamedjavid — APPROVED** [review]
Great work, @ilmotta! 🚀 

Shall we add this in [new-guidelines.md](https://github.com/status-im/status-mobile/blob/develop/doc/new-guidelines.md)?

**Comment by ilmotta** [comment]
> Great work, @ilmotta! rocket
> 
> Shall we add this in [new-guidelines.md](https://github.com/status-im/status-mobile/blob/develop/doc/new-guidelines.md)?

I'm inclined to do that @smohamedjavid. Let me add something :)

Edit: Added to the guidelines as a soft recommendation.

**Review by jo-mut — APPROVED** [review]

**Comment by erikseppanen** [comment]
Q: Are these event handlers actually impure? I looked at the re-frame examples, and the simple definition of data in -> computation -> data out, but I wouldn't construe passing functions in the effects map as impure, because the event handler isn't actually evaluating the function (and therefore not changing the state of the world) - it's passing it as data to something else to evaluate. It will eventually get evaluated when the call is done, but that's another function doing that. I would say that this event handler is actually pure because it is just data in -> computation -> data out. Well, it may not actually be pure because it's a macro, and I haven't looked at what it's doing under-the-hood :)

To be clear, I do approve and support the spirit and intent of this PR! Nice work @ilmotta!

**Review by ajayesivan — APPROVED** [review]

**Review by flexsurfer — APPROVED** [review]

**Review by qfrank — APPROVED** [review]

**Review by OmarBasem — APPROVED** [review]

**Comment by cammellos** [comment]
> 

I think in this case I think @erikseppanen you are correct, they would be pure, since they don't break referential transparency (you can memoize the function and the code behaves exactly the same way), as you correctly point out they don't run any side effect.

You can check the same in haskell, and the generating function would not have to be run in a monad (while the handlers themselves are impure). To give the simplest example (boiling down to the essential):
```function(id) { return function() { console.log(id)}}```
is a pure function, since the output solely depends on `id`, it doesn't not matter that the function returned is impure.

It makes things a bit more opaque to have functions as return value, and I think Icaro's change is much better in terms of explicitness and homogenization of the code, but in terms of purity it's equivalent I believe.

**Review by ibrkhalil — APPROVED** [review]

**Review by rasom — APPROVED** [review]

**Comment by rasom** [comment]
>  and I think Icaro's change is much better in terms of explicitness and homogenization of the code, but in terms of purity it's equivalent I believe.

it is also a bit more testable (if you want to check effect's params for some reason)

**Comment by cammellos** [comment]
> > and I think Icaro's change is much better in terms of explicitness and homogenization of the code, but in terms of purity it's equivalent I believe.
> 
> it is also a bit more testable (if you want to check effect's params for some reason)

oh yes,that 100%

**Comment by ilmotta** [comment]
Thanks for reviewing @erikseppanen, @cammellos, @rasom. I know you guys perfectly understand what I'm going to say, but just to reiterate. re-frame consider such event handlers impure in a _practical_ sense, that's why they say it's only a "90% solution" as I shared the quote.

If the anonymous functions are extracted to vars, then yes, then the event handlers become pure, but that's not the way to go to me, so I discarded this solution. I prefer to fully embrace what re-frame suggests and what's preferred in Clojure in general, just data. Also extracting to vars to force the event handler to be pure still keeps the effects' map more opaque to inspection, which I missed many times during development.

The code below shows that **in practice**, our event handlers are NOT referentially transparent, thus not pure. So theoretically speaking, I agree with you guys, we can *make them pure*, but in practice they're impure 95%+ of the time.

```clojure
(rf/defn mark-as-seen
  {:events [:activity-center/mark-as-seen]}
  [_]
  {:json-rpc/call
   [{:method     "wakuext_markAsSeenActivityCenterNotifications"
     :params     []
     :on-success #(rf/dispatch [:activity-center/mark-as-seen-success])
     :on-error   #(rf/dispatch [:activity-center/mark-as-seen-error])}]})

;; This throws
(assert (= (mark-as-seen {:db {}})
           (mark-as-seen {:db {}})))
```

**Comment by cammellos** [comment]
> 

I get what you are saying, though not sure that it makes the function impure or does not make them referentially transparent. I would consider them both from a theoretically point and a practical point referentially transparent, according to the definition https://en.wikipedia.org/wiki/Expression_(computer_science), since you can replace this by it's value and the behavior of the program does not change, both from a practial  & theoretical standpoint.

Re-frame example is a bit ambigous, since here https://day8.github.io/re-frame/EffectfulHandlers/#another-example it goes from a clear side effect `(GET` to removing it, and also removing the anonymous functions and replacing with vectors, but to me the issue he's pointing out is `GET`, while not sure he's saying #() is not to be considered pure, it's just much nicer/descriptive/testable to use [:event].


It just boils down to the fact that functions are not really comparable for equality, I think that's what we are looking at and what we are discussing, though is a bit tangential to pure/impure/referentially transparent (they are compared I suppose by reference, not by value).

You could make an argument that since allocating memory is a side effect, but that's not generally what we intend for referentially transparent or function purity (unless we explicitly call malloc etc), since it's mostly an execution/runtime concern (nothing is pure at that point as soon as you run it on a machine), but we think of this thing separately from the actual nitty gritty machine execution (unless of course we explicitly handle this, malloc/etc).

But we digress I suppose, it's mostly just tech talk and this is a huge improvement, thanks a lot for the change and  :)

**Comment by ilmotta** [comment]
> I would consider them both from a theoretically point and a practical point referentially transparent, according to the definition https://en.wikipedia.org/wiki/Expression_(computer_science), since you can replace this by it's value and the behavior of the program does not change, both from a practical & theoretical standpoint.

I can't agree with that @cammellos, but as you say, let's not worry too much about such details.

Still, I can't resist just going one step forward in this discussion. From Wikipedia:

> its value must be the same for the same inputs and its evaluation must have no [side effects](https://en.wikipedia.org/wiki/Side_effect_(computer_science)). An expression that is not referentially transparent is called referentially opaque.

We need to satisfy both properties for functions to be pure. Our event handlers have no side-effects, but the returned value is _always_ different. These event handlers are somewhat analogous to a function that generates a random number. I think this is not tangential because the values are indeed different.

Also in Wikipedia:

> An [expression](https://en.wikipedia.org/wiki/Expression_(programming)) is called referentially transparent if it can be [replaced](https://en.wikipedia.org/wiki/Rewriting) with its corresponding value (and vice-versa) without changing the program's behavior.

I can imagine the *behavior* of the system changing because equality checks could return false.

> Re-frame example is a bit ambiguous

I agree, the re-frame docs require some inferring on our part.

> You could make an argument that since allocating memory is a side effect, but that's not generally what we intend for referentially transparent or function purity (unless we explicitly call malloc etc), since it's mostly an execution/runtime concern (nothing is pure at that point as soon as you run it on a machine), but we think of this thing separately from the actual nitty gritty machine execution (unless of course we explicitly handle this, malloc/etc).

:+1: That's super important, there's always a physical effect, so in the end something can always fail and behave differently. We could probably say purity is a concept without any physical meaning.

**Comment by erikseppanen** [comment]
Love the discussion!

> It just boils down to the fact that functions are not really comparable for equality, I think that's what we are looking at and what we are discussing, though is a bit tangential to pure/impure/referentially transparent (they are compared I suppose by reference, not by value).

@cammellos This summarizes how I currently think about it. I think of the anonymous functions in a 'code as data' sense, and ideally (IMO) would be considered on that basis: as data. I'd consider it to be a shortcoming of our tooling that the test can't understand it as such and equate two instances of the same 'code as data' as equivalent.

While trying to write tests for `add-new-contact`, I was similarly frustrated with the same problem (testing event handler callback functions), and took an approach which Icaro and I briefly discussed [in this PR](https://github.com/status-im/status-mobile/pull/15891#discussion_r1194481855). That approach introduces a `dispatch` 'seam' in the application code, which the test can `redef` to return its arguments for equality comparison. Even after this this PR is merged (which I approve and support), I think it can serve as a useful pattern for those cases where it's not a `json-rpc/call` effect that we're testing.

EDIT: if I were to explore that approach again, I might try `redef`ing the `dispatch` call directly and 'dispatch' with the separate `dispatcher` function. Not sure if it would work out, but it would be interesting to try!

**Review by erikseppanen — APPROVED** [review]

**Comment by cammellos** [comment]
> @cammellos This summarizes how I currently think about it. I think of the anonymous functions in a 'code as data' sense, and ideally (IMO) would be considered on that basis: as data. I'd consider it to be a shortcoming of our tooling that the test can't understand it as such and equate two instances of the same 'code as data' as equivalent.

Yes, we are I think discussing what a function is, the two point of views are whether a function is domain/codomain and values, and the other is that is domain/codomain/value + a pointer.

In haskel for example, it does not allow to compare functions (which I don't think it's feasible, since you don't even know if a function will ever finish execution). So there I think it would be referentially transparent.

In clojure, it will compare by reference, which I agree with @ilmotta gives different results, though that's where the disagreement is, and that's where we get in the weeds. 
It seems like context is important, so, how the return value is used.

For example a similar argument could be made by a c sum function, since it would return two distinct pieces of memory, and if we were to compare the two in our program, then it would make the function referentially opaque, which is similar to what's happening in clojure. But then you might argue I guess that `=` is a fundamental operator and that's what should be used to define equality and we should abide to that standard to judge whether a function is referentially opaque or no, and taking a pointer of a function `&` in C is yet another operation.
 I think I reached the bottom of my rabbit hole though :), at least on this PR, super interesting stuff!

**Comment by ilmotta** [comment]
I reached the bottom too @cammellos :milky_way: Very interesting perspectives from you and @erikseppanen! And thanks for the reviews and comments here.

**Comment by erikseppanen** [comment]
OK guys, just one more for the road :)

```clojure
(rf/defn mark-as-seen
  {:events [:activity-center/mark-as-seen]}
  [_]
  {:json-rpc/call
   [{:method     "wakuext_markAsSeenActivityCenterNotifications"
     :params     []
     :on-success '#(rf/dispatch [:activity-center/mark-as-seen-success])
     :on-error   '#(rf/dispatch [:activity-center/mark-as-seen-error])}]})

;; This DOES NOT throw!
(assert (= (mark-as-seen {:db {}})
           (mark-as-seen {:db {}})))
```

**Comment by ilmotta** [comment]
> OK guys, just one more for the road :)
> 
> ```clojure
> (rf/defn mark-as-seen
>   {:events [:activity-center/mark-as-seen]}
>   [_]
>   {:json-rpc/call
>    [{:method     "wakuext_markAsSeenActivityCenterNotifications"
>      :params     []
>      :on-success '#(rf/dispatch [:activity-center/mark-as-seen-success])
>      :on-error   '#(rf/dispatch [:activity-center/mark-as-seen-error])}]})
> 
> ;; This DOES NOT throw!
> (assert (= (mark-as-seen {:db {}})
>            (mark-as-seen {:db {}})))
> ```

The problem with that @erikseppanen is that it's not callable anymore, because it's now considered a list.

```clojure
(type (get-in (mark-as-seen {:db {}}) [:json-rpc/call 0 :on-success]))
;; => cljs.core/List

;; => #object[Function]
;; Throws error, not callable
(let [f (get-in (mark-as-seen {:db {}}) [:json-rpc/call 0 :on-success])]
  (f))
```

**Comment by erikseppanen** [comment]
> > OK guys, just one more for the road :)
> > ```clojure
> > (rf/defn mark-as-seen
> >   {:events [:activity-center/mark-as-seen]}
> >   [_]
> >   {:json-rpc/call
> >    [{:method     "wakuext_markAsSeenActivityCenterNotifications"
> >      :params     []
> >      :on-success '#(rf/dispatch [:activity-center/mark-as-seen-success])
> >      :on-error   '#(rf/dispatch [:activity-center/mark-as-seen-error])}]})
> > 
> > ;; This DOES NOT throw!
> > (assert (= (mark-as-seen {:db {}})
> >            (mark-as-seen {:db {}})))
> > ```
> 
> The problem with that @erikseppanen is that it's not callable anymore, because it's now considered a list.
> 
> ```clojure
> (type (get-in (mark-as-seen {:db {}}) [:json-rpc/call 0 :on-success]))
> ;; => cljs.core/List
> 
> ;; => #object[Function]
> ;; Throws error, not callable
> (let [f (get-in (mark-as-seen {:db {}}) [:json-rpc/call 0 :on-success])]
>   (f))
> ```

Yes, I wasn't thinking of it as a full-fledged solution for this, I don't know if that would be possible. I have done something like this before in clojure, using `eval` to evaluate stored data transformation code snippets, which was pretty cool!

But I added that snippet here because it seemed a good punctuation for the discussion, because `quote` tells clojure to not evaluate the function, and just return it as a data, which might unstick that sticking point regarding anonymous functions. And it illustrates what I was thinking of in terms of 'code as data'.

Glad we had an opportunity to talk about this stuff!

**Comment by ilmotta** [comment]
> > > OK guys, just one more for the road :)
> > > ```clojure
> > > (rf/defn mark-as-seen
> > >   {:events [:activity-center/mark-as-seen]}
> > >   [_]
> > >   {:json-rpc/call
> > >    [{:method     "wakuext_markAsSeenActivityCenterNotifications"
> > >      :params     []
> > >      :on-success '#(rf/dispatch [:activity-center/mark-as-seen-success])
> > >      :on-error   '#(rf/dispatch [:activity-center/mark-as-seen-error])}]})
> > > 
> > > ;; This DOES NOT throw!
> > > (assert (= (mark-as-seen {:db {}})
> > >            (mark-as-seen {:db {}})))
> > > ```
> > 
> > 
> > The problem with that @erikseppanen is that it's not callable anymore, because it's now considered a list.
> > ```clojure
> > (type (get-in (mark-as-seen {:db {}}) [:json-rpc/call 0 :on-success]))
> > ;; => cljs.core/List
> > 
> > ;; => #object[Function]
> > ;; Throws error, not callable
> > (let [f (get-in (mark-as-seen {:db {}}) [:json-rpc/call 0 :on-success])]
> >   (f))
> > ```
> 
> Yes, I wasn't thinking of it as a full-fledged solution for this, I don't know if that would be possible. I have done something like this before in clojure, using `eval` to evaluate stored data transformation code snippets, which was pretty cool!
> 
> But I added that snippet here because it seemed a good punctuation for the discussion, because `quote` tells clojure to not evaluate the function, and just return it as a data, which might unstick that sticking point regarding anonymous functions. And it illustrates what I was thinking of in terms of 'code as data'.
> 
> Glad we had an opportunity to talk about this stuff!

Excellent point @erikseppanen! A good illustration of the power of a Lisp.

**Pull request merged**